php-frontend/src/main/java/org/sonar/php/filters/SuppressWarningFilter.java

```diff
@@ -0,0 +1,124 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.php.filters;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.plugins.php.api.tree.CompilationUnitTree;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.declaration.AttributeGroupTree;
+import org.sonar.plugins.php.api.tree.declaration.AttributeTree;
+import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
+import org.sonar.plugins.php.api.tree.expression.LiteralTree;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxToken;
+import org.sonar.plugins.php.api.tree.lexical.SyntaxTrivia;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+public class SuppressWarningFilter extends PHPVisitorCheck implements FilterPhpIssue {
+
+  private final Map<String, Map<Integer, Set<String>>> suppressedRulesPerLinePerFile = new HashMap<>();
+  private String fileUri;
+
+  private static final String ARGUMENT_FORMAT = "\"[a-zA-Z0-9:]++\"";
+  private static final String ARGUMENTS_FORMAT = ARGUMENT_FORMAT + "(?:\\s*+,\\s*+"+ARGUMENT_FORMAT+")*+";
+  private static final Pattern SUPPRESS_WARNING_COMMENT_PATTERN = Pattern.compile("@SuppressWarnings\\s*+\\(\\s*+(?<arguments>"+ARGUMENTS_FORMAT+")\\s*+\\)");
+
+  public void reset() {
+    suppressedRulesPerLinePerFile.clear();
+  }
+
+  public void scanCompilationUnit(String fileUri, CompilationUnitTree tree) {
+    this.fileUri = fileUri;
+    super.visitCompilationUnit(tree);
+  }
+
+  @Override
+  public boolean accept(String fileName, String ruleName, int line) {
+    return !suppressedRulesPerLinePerFile.getOrDefault(fileName, Collections.emptyMap())
+      .getOrDefault(line, Collections.emptySet()).contains(ruleName);
+  }
+
+  @Override
+  public void visitAttribute(AttributeTree tree) {
+    Set<String> rulesSuppressed = tree.arguments().stream()
+      .map(CallArgumentTree::value)
+      // consider only string literal
+      .filter(expr -> expr.is(Tree.Kind.REGULAR_STRING_LITERAL)).map(LiteralTree.class::cast)
+      .map(LiteralTree::value)
+      .map(SuppressWarningFilter::stripDoubleQuotes)
+      .collect(Collectors.toSet());
+    suppressedRulesPerLinePerFile.computeIfAbsent(fileUri, key -> new HashMap<>()).put(getAttributeLine(tree), rulesSuppressed);
+    super.visitAttribute(tree);
+  }
+
+  public int getAttributeLine(AttributeTree tree) {
+    // TODO: SONARPHP-1368 replace this logic "ignore next line issue" by more advanced "the scope of parent group attribute"
+    AttributeGroupTree parent = (AttributeGroupTree) tree.getParent();
```

Comment:
Make sure that there is no cast error.

Reply:
There cannot be any: from how the parser work, an AttributeTree is necessarily the child of an AttributeGroupTree.
I could also put a defensive check just in case, but then we would have an uncovered test case because it is impossible to have such case (unless we manually build such tree in a very specific test case, which wouldn't make much sense).
I would tend to keep it as it is.

Reply:
If it's not possible that it's fine. I only remembered in my ticket, that running the its uncovered some cast exceptions for such cases.